sdk/azidentity/chained_token_credential.go

```diff
@@ -18,42 +19,45 @@ type ChainedTokenCredential struct {
 
 // NewChainedTokenCredential creates an instance of ChainedTokenCredential with the specified TokenCredential sources.
 func NewChainedTokenCredential(sources ...azcore.TokenCredential) (*ChainedTokenCredential, error) {
-	if len(sources) == 0 {
-		return nil, &CredentialUnavailableError{CredentialType: "Chained Token Credential", Message: "Length of sources cannot be 0"}
-	}
-	for _, source := range sources {
-		if source == nil {
-			return nil, &CredentialUnavailableError{CredentialType: "Chained Token Credential", Message: "Sources cannot contain a nil TokenCredential"}
-		}
-	}
 	return &ChainedTokenCredential{sources: sources}, nil
 }
 
 // GetToken sequentially calls TokenCredential.GetToken on all the specified sources, returning the first non default AccessToken.
-func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
-	var token *azcore.AccessToken
-	var err error
+func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (token *azcore.AccessToken, err error) {
 	var errList []*CredentialUnavailableError
-	for i := 0; i < len(c.sources); i++ {
+	for _, cred := range c.sources { // loop through all of the credentials provided in sources
+		token, err = cred.GetToken(ctx, opts) // make a GetToken request for the current credential in the loop
 		var credErr *CredentialUnavailableError
-		token, err = c.sources[i].GetToken(ctx, opts)
-		if errors.As(err, &credErr) {
-			errList = append(errList, credErr)
-		} else if err != nil {
-			errList = append(errList, &CredentialUnavailableError{CredentialType: "Chained Token Credential", Message: err.Error()})
-			break
+		if errors.As(err, &credErr) { // check if we received a CredentialUnavailableError
+			errList = append(errList, credErr) // if we did receive a CredentialUnavailableError then we append it to our error slice and continue looping for a good credential
+		} else if err != nil { // if we receive some other type of error then we must stop looping and process the error accordingly
+			var authenticationFailed *AuthenticationFailedError
+			if errors.As(err, &authenticationFailed) { // if the error is an AuthenticationFailedError we return the error related to the invalid credential and append all of the other error messages received prior to this point
+				return nil, &AuthenticationFailedError{msg: "Received an AuthenticationFailedError, there is an invalid credential in the chain. " + createChainedErrorMessage(errList), inner: err}
+			}
+			return nil, fmt.Errorf("Received an unexpected error: %w", err) // if we receive some other error type this is unexpected and we simple return the unexpected error
 		} else {
-			return token, nil
+			return token, nil // if we did not receive an error then we return the token
 		}
 	}
 	// This condition should never be true
 	if token == nil && len(errList) == 0 {
```

Comment:
Is this actually true now, that it can't happen?  Since we no longer ensure `len(sources) > 0` it seems you can now hit this case if `NewChainedTokenCredential()` is called without any parameters.

Reply:
Also true, I'm updating the comment to explain why this could happen now

Reply:
I reverted the changes so I also reverted the comment. This condition should never be true now

Reply:
Can this block just be removed then?

Reply:
Yes